File: src/main/java/io/enthusia/express/gui/ShippingService.java
package io.enthusia.express.gui;

import io.enthusia.express.db.MailRepository;
import io.enthusia.express.hook.CombatLogXHook;
import io.enthusia.express.util.ContainerScanner;
import io.enthusia.express.util.ItemCodec;
import io.enthusia.express.util.Text;
import io.enthusia.express.util.SoundFeedback;
import io.enthusia.express.util.SoundFeedback.Cue;
import java.util.*;
import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.OfflinePlayer;
import org.bukkit.entity.Player;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.persistence.PersistentDataType;
import org.bukkit.NamespacedKey;
import org.bukkit.plugin.java.JavaPlugin;

public final class ShippingService {
  public static final String TITLE_PREFIX = "Enthusia Express: Ship to ";
  public static final int PACKAGE_SLOT = 13;
  public static final int CONFIRM_SLOT = 15;
  public static final int CANCEL_SLOT = 11;
  private final io.enthusia.express.util.MainThread main;
  private final Map<UUID, Inventory> inventories = new HashMap<>();
  private final Set<UUID> pending = new HashSet<>();
  private final JavaPlugin plugin;
  private final MailRepository repository;
  private final CombatLogXHook combatHook;
  private final Map<UUID, UUID> targets = new HashMap<>();
  private final SoundFeedback sounds;
  private final NamespacedKey placeholderKey;

  public ShippingService(
      JavaPlugin plugin,
      MailRepository repository,
      CombatLogXHook combatHook,
      io.enthusia.express.util.MainThread main) {
    this(plugin, repository, combatHook, main, new SoundFeedback(plugin));
  }

  public ShippingService(
      JavaPlugin plugin,
      MailRepository repository,
      CombatLogXHook combatHook,
      io.enthusia.express.util.MainThread main,
      SoundFeedback sounds) {
    this.main = main;
    this.plugin = plugin;
    this.repository = repository;
    this.combatHook = combatHook;
    this.sounds = sounds;
    this.placeholderKey = new NamespacedKey(plugin, "shipping-placeholder");
  }

  public void open(Player sender, OfflinePlayer target) {
    if (!sender.hasPermission("enthusiaexpress.use")
        || !sender.hasPermission("enthusiaexpress.packages.send")) {
      sender.sendMessage(Text.msg(plugin.getConfig(), "no-permission"));
      return;
    }
    if (pending.contains(sender.getUniqueId())) {
      sender.sendMessage("\u00a7eYour shipment is still being saved.");
      return;
    }
    if (!combatHook.mayUseMail(sender)) {
      sender.sendMessage(
          Text.msg(
              plugin.getConfig(),
              combatHook.isAvailable() ? "combat-blocked" : "combatlogx-missing"));
      return;
    }
    Player online = target.getPlayer();
    if (online != null && online.isOnline()) {
      sender.sendMessage(Text.msg(plugin.getConfig(), "target-online"));
      return;
    }
    sender.closeInventory();
    targets.put(sender.getUniqueId(), target.getUniqueId());
    Inventory inv = Bukkit.createInventory(null, 27, TITLE_PREFIX + target.getName());
    inv.setItem(CANCEL_SLOT, button(Material.BARRIER, "\u00a7cCancel"));
    inv.setItem(CONFIRM_SLOT, button(Material.LIME_CONCRETE, "\u00a7aConfirm shipment"));
    refreshPlaceholder(inv);
    inventories.put(sender.getUniqueId(), inv);
    sender.openInventory(inv);
  }

  public boolean owns(Player player, Inventory inventory) {
    return inventory != null && inventories.get(player.getUniqueId()) == inventory;
  }

  public void defer(Player player, Inventory inventory, boolean confirm) {
    Bukkit.getScheduler()
        .runTask(
            plugin,
            () -> {
              if (!owns(player, inventory)
                  || player.getOpenInventory().getTopInventory() != inventory) return;
              if (confirm) confirm(player, inventory);
              else cancel(player);
            });
  }

  public void cancel(Player player) {
    player.closeInventory();
  }

  public void confirm(Player sender, Inventory inv) {
    if (!sender.hasPermission("enthusiaexpress.use")
        || !sender.hasPermission("enthusiaexpress.packages.send")) {
      sender.sendMessage(Text.msg(plugin.getConfig(), "no-permission"));
      return;
    }
    if (pending.contains(sender.getUniqueId())) {
      sender.sendMessage("\u00a7eYour shipment is still being saved.");
      return;
    }
    if (!combatHook.mayUseMail(sender)) {
      sender.sendMessage(
          Text.msg(
              plugin.getConfig(),
              combatHook.isAvailable() ? "combat-blocked" : "combatlogx-missing"));
      sender.closeInventory();
      return;
    }
    if (!owns(sender, inv)) return;
    UUID targetId = targets.get(sender.getUniqueId());
    if (targetId == null) return;
    OfflinePlayer target = Bukkit.getOfflinePlayer(targetId);
    if (target.isOnline()) {
      sender.sendMessage(Text.msg(plugin.getConfig(), "target-online"));
      sender.closeInventory();
      return;
    }
    ItemStack packageItem = inv.getItem(PACKAGE_SLOT);
    if (isPlaceholder(packageItem)) packageItem = null;
    if (!ContainerScanner.isAllowedShippingContainer(packageItem)) {
      sender.sendMessage(Text.msg(plugin.getConfig(), "invalid-container"));
      return;
    }
    if (packageItem.getAmount() != 1) {
      sender.sendMessage("\u00a7cSend one container at a time.");
      return;
    }
    int count;
    int cost;
    try {
      count =
          ContainerScanner.countPackedItems(
              packageItem, plugin.getConfig().getInt("mail.max-recursive-container-depth", 8));
      cost = Math.multiplyExact(count, plugin.getConfig().getInt("mail.raw-gold-per-item", 1));
    } catch (IllegalArgumentException | ArithmeticException e) {
      sender.sendMessage("\u00a7cContainer nesting or shipment cost exceeds the configured limits.");
      return;
    }
    if (count <= 0) {
      sender.sendMessage(Text.msg(plugin.getConfig(), "empty-container"));
      return;
    }
    int have = countMaterial(sender, Material.RAW_GOLD);
    if (have < cost) {
      sender.sendMessage(
          Text.msg(
              plugin.getConfig(),
              "insufficient-gold",
              Map.of("cost", String.valueOf(cost), "have", String.valueOf(have))));
      return;
    }
    // Copy + encode on the primary thread before database work.
    ItemStack payloadItem = packageItem.clone();
    byte[] payload;
    try {
      payload = ItemCodec.encode(payloadItem);
    } catch (RuntimeException e) {
      sender.sendMessage("\u00a7cCould not encode that container.");
      return;
    }
    inv.setItem(PACKAGE_SLOT, null);
    removeMaterial(sender, Material.RAW_GOLD, cost);
    sender.closeInventory();
    targets.remove(sender.getUniqueId());
    String targetName = Optional.ofNullable(target.getName()).orElse(targetId.toString());
    pending.add(sender.getUniqueId());
    main.complete(
        repository.insertMailLimited(
            sender.getUniqueId(),
            sender.getName(),
            targetId,
            targetName,
            io.enthusia.express.mail.MailType.PACKAGE,
            payload,
            count,
            plugin
                .getConfig()
                .getBoolean("mail.limits.one-outstanding-package-per-recipient", false)),
        (result, error) -> {
          pending.remove(sender.getUniqueId());
          if (error != null) {
            // Compensate on the server thread; use the current session after a reconnect.
            refundPlayer(sender, payloadItem, cost);
            sender.sendMessage("\u00a7cShipment failed; your package and fee were refunded.");
            plugin.getLogger().severe("Package insert failed: " + error.getMessage());
          } else if (result.isEmpty()) {
            refundPlayer(sender, payloadItem, cost);
            sender.sendMessage(Text.msg(plugin.getConfig(), "outstanding-package"));
          } else {
            sender.sendMessage(
                Text.msg(
                    plugin.getConfig(),
                    "package-sent",
                    Map.of(
                        "target",
                        targetName,
                        "cost",
                        String.valueOf(cost),
                        "items",
                        String.valueOf(count))));
            sounds.play(sender, Cue.PACKAGE_SEND);
          }
        });
  }

  public void returnPackageOnClose(Player player, Inventory inv) {
    if (!owns(player, inv)) return;
    ItemStack stack = inv.getItem(PACKAGE_SLOT);
    if (stack != null && !isPlaceholder(stack) && !stack.getType().isAir()) {
      inv.setItem(PACKAGE_SLOT, null);
      Map<Integer, ItemStack> overflow = player.getInventory().addItem(stack);
      overflow
          .values()
          .forEach(item -> player.getWorld().dropItemNaturally(player.getLocation(), item));
    }
    targets.remove(player.getUniqueId());
    inventories.remove(player.getUniqueId());
  }

  public void deferPlaceholderRefresh(Player player, Inventory inventory) {
    Bukkit.getScheduler()
        .runTask(
            plugin,
            () -> {
              if (owns(player, inventory)) refreshPlaceholder(inventory);
            });
  }

  public void refreshPlaceholder(Inventory inventory) {
    ItemStack current = inventory.getItem(PACKAGE_SLOT);
    if (current == null || current.getType().isAir()) inventory.setItem(PACKAGE_SLOT, placeholder());
  }

  public boolean isPlaceholder(ItemStack item) {
    if (item == null || item.getType() != Material.GRAY_STAINED_GLASS_PANE) return false;
    ItemMeta meta = item.getItemMeta();
    return meta != null
        && meta.getPersistentDataContainer().has(placeholderKey, PersistentDataType.BYTE);
  }

  public void shutdown() {
    for (Player player : Bukkit.getOnlinePlayers())
      if (inventories.containsKey(player.getUniqueId())) player.closeInventory();
  }

  private static void refundPlayer(Player sender, ItemStack payloadItem, int cost) {
    Player refundTarget =
        Optional.ofNullable(Bukkit.getPlayer(sender.getUniqueId())).orElse(sender);
    give(refundTarget, payloadItem);
    for (int remaining = cost; remaining > 0; remaining -= Math.min(64, remaining))
      give(refundTarget, new ItemStack(Material.RAW_GOLD, Math.min(64, remaining)));
    if (!refundTarget.isOnline()) refundTarget.saveData();
  }

  private static void give(Player player, ItemStack item) {
    player
        .getInventory()
        .addItem(item)
        .values()
        .forEach(left -> player.getWorld().dropItemNaturally(player.getLocation(), left));
  }

  private static ItemStack button(Material material, String name) {
    ItemStack stack = new ItemStack(material);
    ItemMeta meta = stack.getItemMeta();
    meta.setDisplayName(name);
    stack.setItemMeta(meta);
    return stack;
  }

  private ItemStack placeholder() {
    ItemStack stack = new ItemStack(Material.GRAY_STAINED_GLASS_PANE);
    ItemMeta meta = stack.getItemMeta();
    meta.setDisplayName("\u00a77Place package here");
    meta.setLore(List.of("\u00a77Accepted: Shulker Boxes and Bundles"));
    meta.getPersistentDataContainer().set(placeholderKey, PersistentDataType.BYTE, (byte) 1);
    stack.setItemMeta(meta);
    return stack;
  }

  private static int countMaterial(Player player, Material material) {
    int count = 0;
    for (ItemStack item : player.getInventory().getStorageContents()) {
      if (item != null && item.getType() == material) count += item.getAmount();
    }
    return count;
  }

  private static void removeMaterial(Player player, Material material, int amount) {
    int remaining = amount;
    ItemStack[] contents = player.getInventory().getStorageContents();
    for (int i = 0; i < contents.length && remaining > 0; i++) {
      ItemStack item = contents[i];
      if (item == null || item.getType() != material) continue;
      int take = Math.min(item.getAmount(), remaining);
      item.setAmount(item.getAmount() - take);
      if (item.getAmount() <= 0) contents[i] = null;
      remaining -= take;
    }
    player.getInventory().setStorageContents(contents);
  }
}

File: src/main/kotlin/io/enthusia/express/infrastructure/gui/ShippingService.kt
// Private callback contracts document thread ownership and recovery; CodeRabbit requires method documentation.
@file:Suppress("CommentOverPrivateFunction")

package io.enthusia.express.infrastructure.gui

import io.enthusia.express.application.MailStore
import io.enthusia.express.domain.MailType
import io.enthusia.express.infrastructure.hook.CombatLogXHook
import io.enthusia.express.infrastructure.payment.PaymentReceipt
import io.enthusia.express.infrastructure.payment.ShippingPayments
import io.enthusia.express.infrastructure.util.ContainerScanner
import io.enthusia.express.infrastructure.util.ItemCodec
import io.enthusia.express.infrastructure.util.MainThread
import io.enthusia.express.infrastructure.util.SoundFeedback
import io.enthusia.express.infrastructure.util.Text
import java.util.UUID
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.OfflinePlayer
import org.bukkit.entity.Player
import org.bukkit.inventory.Inventory
import org.bukkit.inventory.ItemStack
import org.bukkit.persistence.PersistentDataType
import org.bukkit.plugin.java.JavaPlugin

// This lifecycle owner keeps inventory identity, pending payments and close compensation together.
@Suppress("TooManyFunctions")
class ShippingService @JvmOverloads constructor(
    private val plugin: JavaPlugin,
    private val repository: MailStore,
    private val combatHook: CombatLogXHook,
    private val main: MainThread,
    private val sounds: SoundFeedback = SoundFeedback(plugin),
) {
    private val payments = ShippingPayments(plugin)
    private val placeholderKey = NamespacedKey(plugin, "shipping-placeholder")
    private val inventories = HashMap<UUID, Inventory>()
    private val pending = HashSet<UUID>()
    private val targets = HashMap<UUID, UUID>()

    /** Centralize permission, in-flight-send and combat checks before accepting shipping actions. */
    private fun validateShippingAccess(sender: Player, closeBlocked: Boolean): Boolean {
        if (!sender.hasPermission("enthusiaexpress.use") || !sender.hasPermission("enthusiaexpress.packages.send")) {
            sender.sendMessage(Text.msg(plugin.config, "no-permission"))
            return false
        }
        if (pending.contains(sender.uniqueId)) {
            sender.sendMessage("§eYour shipment is still being saved.")
            return false
        }
        if (!combatHook.mayUseMail(sender)) {
            sender.sendMessage(Text.msg(plugin.config, if (combatHook.isAvailable()) "combat-blocked" else "combatlogx-missing"))
            if (closeBlocked) sender.closeInventory()
            return false
        }
        return true
    }

    /** Create the sender-owned cargo menu for an offline recipient. */
    fun open(sender: Player, target: OfflinePlayer) {
        if (!validateShippingAccess(sender, false)) return
        val online = target.player
        if (online != null && online.isOnline) {
            sender.sendMessage(Text.msg(plugin.config, "target-online"))
            return
        }
        sender.closeInventory()
        targets[sender.uniqueId] = target.uniqueId
        val inv = Bukkit.createInventory(null, 27, TITLE_PREFIX + target.name)
        inv.setItem(CANCEL_SLOT, button(Material.BARRIER, "§cCancel"))
        inv.setItem(CONFIRM_SLOT, button(Material.LIME_CONCRETE, "§aConfirm shipment"))
        refreshPlaceholder(inv)
        inventories[sender.uniqueId] = inv
        sender.openInventory(inv)
    }

    /** Compare the exact inventory instance with the sender-owned shipping menu. */
    fun owns(player: Player, inventory: Inventory?): Boolean = inventory != null && inventories[player.uniqueId] === inventory

    /** Move confirm or cancel handling out of the click event and reject stale inventories. */
    fun defer(player: Player, inventory: Inventory, confirm: Boolean) {
        Bukkit.getScheduler().runTask(plugin, Runnable {
            if (owns(player, inventory) && player.openInventory.topInventory === inventory) {
                if (confirm) confirm(player, inventory) else cancel(player)
            }
        })
    }

    /** Close the shipping menu so its close handler returns any cargo. */
    fun cancel(player: Player) { player.closeInventory() }

    /** Revalidate sender and recipient state, then prepare and submit the current cargo. */
    fun confirm(sender: Player, inv: Inventory) {
        if (!validateShippingAccess(sender, true)) return
        if (!owns(sender, inv)) return
        val targetId = targets[sender.uniqueId] ?: return
        val target = Bukkit.getOfflinePlayer(targetId)
        if (target.isOnline) {
            sender.sendMessage(Text.msg(plugin.config, "target-online"))
            sender.closeInventory()
            return
        }
        val shipment = prepareShipment(sender, inv) ?: return
        chargeAndSubmit(sender, inv, target, shipment)
    }

    private data class PreparedShipment(val payloadItem: ItemStack, val payload: ByteArray, val count: Int, val cost: Int)

    /** Validate nesting and payload limits before calculating postage or removing items. */
    // Item serializers are supplied by Paper and may reject malformed/version-specific metadata.
    @Suppress("TooGenericExceptionCaught")
    private fun prepareShipment(sender: Player, inv: Inventory): PreparedShipment? {
        val packageItem = inv.getItem(PACKAGE_SLOT)
        if (isPlaceholder(packageItem) || !ContainerScanner.isAllowedShippingContainer(packageItem)) {
            sender.sendMessage(Text.msg(plugin.config, "invalid-container"))
            return null
        }
        checkNotNull(packageItem)
        if (packageItem.amount != 1) {
            sender.sendMessage("§cSend one container at a time.")
            return null
        }
        val count: Int
        val cost: Int
        try {
            count = ContainerScanner.countPackedItems(packageItem, plugin.config.getInt("mail.max-recursive-container-depth", 8))
            cost = Math.multiplyExact(count, plugin.config.getInt("mail.raw-gold-per-item", 1))
        } catch (e: IllegalArgumentException) {
            sender.sendMessage("§cContainer nesting or shipment cost exceeds the configured limits.")
            return null
        } catch (e: ArithmeticException) {
            sender.sendMessage("§cContainer nesting or shipment cost exceeds the configured limits.")
            return null
        }
        if (count <= 0) {
            sender.sendMessage(Text.msg(plugin.config, "empty-container"))
            return null
        }
        // Copy and encode on the primary thread before database work.
        val payloadItem = packageItem.clone()
        val payload: ByteArray
        try {
            payload = ItemCodec.encode(payloadItem)
        } catch (e: RuntimeException) {
            plugin.logger.log(java.util.logging.Level.WARNING, "Cannot serialize package for ${sender.uniqueId}", e)
            sender.sendMessage("§cCould not encode that container.")
            return null
        }
        return PreparedShipment(payloadItem, payload, count, cost)
    }

    /** Charge one payment route, reserve cargo and compensate rejected or failed persistence. */
    private fun chargeAndSubmit(sender: Player, inv: Inventory, target: OfflinePlayer, shipment: PreparedShipment) {
        val payment = payments.charge(sender, shipment.cost)
        val receipt = payment.receipt
        if (receipt == null) {
            sender.sendMessage(Text.msg(plugin.config,
                paymentFailureMessage(payment),
                mapOf("cost" to shipment.cost.toString(), "have" to java.math.BigDecimal.valueOf(payment.balance).stripTrailingZeros().toPlainString())))
            return
        }
        inv.setItem(PACKAGE_SLOT, null)
        sender.closeInventory()
        targets.remove(sender.uniqueId)
        val targetName = target.name ?: target.uniqueId.toString()
        pending.add(sender.uniqueId)
        main.complete(repository.insertMailLimited(sender.uniqueId, sender.name, target.uniqueId, targetName,
            MailType.PACKAGE, shipment.payload, shipment.count, plugin.config.getBoolean("mail.limits.one-outstanding-package-per-recipient", false))) { result, error ->
            pending.remove(sender.uniqueId)
            if (error != null) {
                if (refundPlayer(sender, shipment.payloadItem, receipt))
                    sender.sendMessage("§cShipment failed; your package and fee were refunded.")
                plugin.logger.severe("Package insert failed: " + error.message)
            } else if (result!!.isEmpty) {
                if (refundPlayer(sender, shipment.payloadItem, receipt))
                    sender.sendMessage(Text.msg(plugin.config, "outstanding-package"))
            } else {
                sounds.play(sender, SoundFeedback.Cue.PACKAGE_SEND)
                sender.sendMessage(Text.msg(plugin.config, "package-sent", mapOf("target" to targetName, "cost" to shipment.cost.toString(), "items" to shipment.count.toString())))
            }
        }
    }

    /** Choose unavailable, physical-gold or currency messaging from the actual payment result. */
    private fun paymentFailureMessage(payment: io.enthusia.express.infrastructure.payment.ChargeResult): String = when {
        payment.unavailable -> "payment-unavailable"
        payment.source == io.enthusia.express.infrastructure.payment.PaymentSource.CURRENCY -> "insufficient-currency"
        else -> "insufficient-gold"
    }

    /** Remove and return cargo during the close event, never returning the placement marker. */
    fun returnPackageOnClose(player: Player, inv: Inventory) {
        if (!owns(player, inv)) return
        val stack = inv.getItem(PACKAGE_SLOT)
        inv.setItem(PACKAGE_SLOT, null)
        if (stack != null && !isPlaceholder(stack) && !stack.type.isAir) {
            player.inventory.addItem(stack).values.forEach { player.world.dropItemNaturally(player.location, it) }
        }
        targets.remove(player.uniqueId)
        inventories.remove(player.uniqueId)
    }

    /** Recognize the tagged gray placement marker instead of ordinary player cargo. */
    fun isPlaceholder(item: ItemStack?): Boolean = item?.type == Material.GRAY_STAINED_GLASS_PANE &&
        item.itemMeta?.persistentDataContainer?.has(placeholderKey, PersistentDataType.BYTE) == true

    /** Show placement guidance only when the cargo slot is empty. */
    fun refreshPlaceholder(inventory: Inventory) {
        val current = inventory.getItem(PACKAGE_SLOT)
        if (current != null && !current.type.isAir) return
        val marker = button(Material.GRAY_STAINED_GLASS_PANE, "§7Place package here")
        val meta = marker.itemMeta!!
        meta.lore = listOf("§7Accepted: Shulker Boxes and Bundles")
        meta.persistentDataContainer.set(placeholderKey, PersistentDataType.BYTE, 1.toByte())
        marker.itemMeta = meta
        inventory.setItem(PACKAGE_SLOT, marker)
    }

    /** Refresh placement guidance next tick only for the same shipping session. */
    fun deferPlaceholderRefresh(player: Player, inventory: Inventory) {
        Bukkit.getScheduler().runTask(plugin, Runnable {
            if (owns(player, inventory)) refreshPlaceholder(inventory)
        })
    }

    /** Defer cursor-to-cargo placement until the inventory click event has completed. */
    fun deferPlaceholderDeposit(player: Player, inventory: Inventory) {
        Bukkit.getScheduler().runTask(plugin, Runnable { depositCursor(player, inventory) })
    }

    /** Transfer a nonempty cursor into the marker slot only while the same menu still owns it. */
    private fun depositCursor(player: Player, inventory: Inventory) {
        if (!player.isOnline || !owns(player, inventory)) return
        if (player.openInventory.topInventory !== inventory || !isPlaceholder(inventory.getItem(PACKAGE_SLOT))) return
        val cursor: ItemStack? = player.itemOnCursor
        if (cursor == null || cursor.type.isAir || isPlaceholder(cursor)) return
        val cargo = cursor.clone()
        player.setItemOnCursor(null)
        inventory.setItem(PACKAGE_SLOT, cargo)
    }

    /** Return cargo and refund its original payment receipt, reporting refused refunds to the operator. */
    private fun refundPlayer(sender: Player, payloadItem: ItemStack, receipt: PaymentReceipt): Boolean {
        val target = Bukkit.getPlayer(sender.uniqueId) ?: sender
        give(target, payloadItem)
        if (!target.isOnline) target.saveData()
        val refunded = receipt.refund()
        if (!refunded) {
            target.sendMessage("§cYour package was returned, but the fee refund failed. Contact an administrator.")
            plugin.logger.severe("Postage refund requires administrator action for ${sender.uniqueId}")
        }
        return refunded
    }

    /** Close owned shipping menus so cargo returns before pending persistence callbacks drain. */
    fun shutdown() {
        for (player in Bukkit.getOnlinePlayers()) if (inventories.containsKey(player.uniqueId)) player.closeInventory()
    }

    companion object {
        const val TITLE_PREFIX = "Enthusia Express: Ship to "
        const val PACKAGE_SLOT = 13
        const val CONFIRM_SLOT = 15
        const val CANCEL_SLOT = 11

        /** Return an item to player storage and drop only inventory overflow. */
        private fun give(player: Player, item: ItemStack) {
            player.inventory.addItem(item).values.forEach { player.world.dropItemNaturally(player.location, it) }
        }

        /** Create a named decorative shipping control. */
        private fun button(material: Material, name: String): ItemStack {
            val stack = ItemStack(material)
            val meta = stack.itemMeta!!
            meta.setDisplayName(name)
            stack.itemMeta = meta
            return stack
        }

    }
}
